Handle isWaitingOnBankAccount for IOU requests

nkuoch opened the pull request:
- [x] HOLD on https://github.com/Expensify/Auth/pull/8131 and https://github.com/Expensify/Web-Expensify/pull/37841
- [x] HOLD on https://github.com/Expensify/Web-Expensify/pull/38042 and https://github.com/Expensify/Auth/pull/8226

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/270705
$ https://github.com/Expensify/App/issues/22763

### Tests
Open browser 1:
Create user admin@thedomain20821.com
Create new workspace and invite employee2@thedomain20821.com
Create workspace OPEN vbba [using test credentials](https://stackoverflowteams.com/c/expensify/questions/342/525#525)

Open browser 2:
Sign into employee2@thedomain20821.com
In the workspace chat, request money
<img width="1193" alt="image" src="https://github.com/Expensify/App/assets/2463975/5cfbb4ff-f248-4f4a-a77b-28ca3c1849d4">

Chat view:
<img width="1196" alt="image" src="https://github.com/Expensify/App/assets/2463975/004fe735-b54c-49f5-9e96-120be3bf637a">

On browser 1 (admin@thedomain20821.com), you should see the request

Chat view:
<img width="1728" alt="image" src="https://github.com/Expensify/App/assets/2463975/2977d042-54db-40b7-8e42-106518da0afb">

Request view:
<img width="1728" alt="image" src="https://github.com/Expensify/App/assets/2463975/569ac3be-9642-490d-a39d-446009acd7e6">

Click on **Pay with Expensify**.
You should see "started settling up, payment is held until ... adds a bank account":

Request view:
<img width="1728" alt="image" src="https://github.com/Expensify/App/assets/2463975/44c2576f-05b2-494e-8f44-997d049a0615">

On browser 2 (employee2@thedomain20821.com)
You should see a green dot and a button to add your bank account.

<img width="1196" alt="image" src="https://github.com/Expensify/App/assets/2463975/30e370da-56f4-44d9-b19d-ce06e9ffef29">

Click on the button and add a personal bank account. You should see the payment as settled:

Request view:
<img width="1194" alt="image" src="https://github.com/Expensify/App/assets/2463975/11fe4f60-2208-4708-958f-329f43086a74">

On browser 1 (admin@thedomain20821.com), make sure you see the request as Paid

Chat view:
<img width="1728" alt="image" src="https://github.com/Expensify/App/assets/2463975/7837189a-4e76-4be5-bd69-283a8ec2bdeb">

Request view:
<img width="1726" alt="image" src="https://github.com/Expensify/App/assets/2463975/09ee3c8c-c11b-4aee-a219-a11f123220ea">

### Offline tests
N/A

### QA Steps
Same as Tests

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message: https://expensify.slack.com/archives/C21FRDWCV/p1687808332892129
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos

The screenshots and videos can be seen in the comments of the PR. Also the reviewer included all the platforms they could run at the time. Checking off the checkbox.

<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

https://user-images.githubusercontent.com/36083550/253777608-39e9fdfe-0bdf-4671-94ac-7f8b6a5d26fa.png

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

</details>

nkuoch committed:
Handle isWaitingForBankAccount when paying request

nkuoch committed:
Handle reimbursed action

nkuoch commented on src/languages/es.js:
```diff
@@ -346,6 +346,7 @@ export default {
         sendMoney: 'Enviar dinero',
         pay: 'Pagar',
         viewDetails: 'Ver detalles',
+        pending: 'Pendiente',
```
@rocio do you mind checking these es translations?

mountiny replied:
I think there is this SO for the translations https://stackoverflowteams.com/c/expensify/questions/10316/10317#10317

nkuoch committed:
Some cleaning

mountiny commented on src/languages/en.js:
```diff
@@ -360,6 +361,9 @@ export default {
         payerOwesAmount: ({payer, amount}) => `${payer} owes ${amount}`,
         payerPaidAmount: ({payer, amount}) => `${payer} paid ${amount}`,
         payerSettled: ({amount}) => `paid ${amount}`,
+        payerSettledUp: ({amount}) => `settled up ${amount}`,
```
This one should be `paid`, we have changed it since the the design doc, so you can use the copy above.

mountiny commented on src/libs/ReportUtils.js:
```diff
@@ -1811,20 +1817,16 @@ function isUnreadWithMention(report) {
  * @returns {boolean}
  */
 function hasOutstandingIOU(report, iouReports) {
-    if (!report || !report.iouReportID || _.isUndefined(report.hasOutstandingIOU)) {
-        return false;
-    }
```
Is it safe to remove this?

nkuoch replied:
It's only used in 1 place: [shouldReportBeInOptionList](https://github.com/Expensify/App/blob/c370ea925847dcf68eb48df0ed984eb8e4112ee1/src/libs/ReportUtils.js#L1930). I wanted to force displaying the iou request on the LHN when IOU is outstanding or is waiting on bank account, but It actually didn't work, so I might as well revert it to how it was. And maybe we do it on purpose to show the IOU details on the LHN only once you click on the message from the parent report? ... but then we'd need to show the red dot on the parent, no? .. I think I'm a bit confused on what we should expect on the LHN.

mountiny replied:
There is a incomplete implementation yet where the expense/iou reports are not showing from the GetChats call so when you log out and log back in you have to click on the report preview to load it

mountiny commented on src/components/MoneyRequestHeader.js:
```diff
@@ -87,9 +87,16 @@ function MoneyRequestHeader(props) {
+    if (isSettled) {
+        description += ` • ${props.translate('iou.settledExpensify')}`;
+    } else if (props.report.isWaitingOnBankAccount) {
+        description += ` • Waiting for credit account`;
```
This should be translated as well I imagine

mountiny commented on src/pages/home/report/ReportActionItemReimbursed.js:
```diff
@@ -0,0 +1,27 @@
+    ...withLocalizePropTypes,
+};
+
+function ReportActionItemReimbursed(props) {
+    return (
+        <Text style={[styles.chatItemMessage, styles.colorMuted]}>
+            {props.translate('iou.payerSettledUp', {amount: props.amount})}{' '}
+            {props.isFromSubmitterAddingBankAccount ? props.translate('iou.afterAddedBankAccount', {submitterDisplayName: props.submitterDisplayName}) : ''}
+        </Text>
+    );
+}
```
I see that this is coming from the design doc but I am curious if we should have two actions for the "Paid $40" logic. cc @luacmartins what do you think?

nkuoch replied:
https://docs.google.com/document/d/17jFcMsAotXvsMhXivoF3u7fyzeyXkKpRGXR88Ud0oWY/edit?disco=AAAAzP95sRQ

luacmartins commented:
@nkuoch is this PR ready for review?

nkuoch commented:
@luacmartins yes for a first pass of code review. Meanwhile, I'll add tests.

nkuoch committed:
Add translations

nkuoch committed:
New copies

nkuoch committed:
Merge branch 'main' into nat-iouw

nkuoch committed:
js style

nkuoch committed:
Merge branch 'main' into nat-iouw

nkuoch committed:
Use getMoneyRequestTotal for displayed amount

nkuoch committed:
Fix console error by adding report billing state

nkuoch committed:
Fix eslint

nkuoch committed:
Some refactoring

nkuoch committed:
Use button instead of link to add a bank account

luacmartins commented on src/pages/home/report/ReportActionItem.js:
```diff
@@ -267,6 +271,24 @@ function ReportActionItem(props) {
                     isHovered={hovered}
                 />
             );
+        } else if (props.action.actionName === CONST.REPORT.ACTIONS.TYPE.REIMBURSEMENTQUEUED) {
+            const submitterDisplayName = PersonalDetailsUtils.getDisplayNameOrDefault(props.personalDetailsList, [props.report.ownerAccountID, 'displayName'], props.report.ownerEmail);
+            children = (
+                <ReportActionItemReimbursementQueued
+                    submitterDisplayName={submitterDisplayName}
+                    isCurrentUserSubmitter={ReportUtils.isCurrentUserSubmitter(props.report.reportID)}
+                />
+            );
+        } else if (props.action.actionName === CONST.REPORT.ACTIONS.TYPE.REIMBURSED) {
+            const totalAmount = ReportUtils.getMoneyRequestTotal(props.report);
+            const submitterDisplayName = PersonalDetailsUtils.getDisplayNameOrDefault(props.personalDetailsList, [props.report.ownerAccountID, 'displayName'], props.report.ownerEmail);
+            children = (
+                <ReportActionItemReimbursed
+                    amount={CurrencyUtils.convertToDisplayString(totalAmount, props.report.currency)}
+                    submitterDisplayName={submitterDisplayName}
+                    isFromSubmitterAddingBankAccount={props.action.originalMessage.isSubmitterAddingBankAccount}
+                />
+            );
```
I wonder if we can just use [ReportActionItemMessage](https://github.com/Expensify/App/blob/27e9eaa99b8c667fdb05bc672c550c5c67f1d55b/src/pages/home/report/ReportActionItemMessage.js). We already use it for a few system messages, e.g [here](https://github.com/Expensify/App/blob/27e9eaa99b8c667fdb05bc672c550c5c67f1d55b/src/pages/home/report/ReportActionItem.js#L296-L301)

luacmartins replied:
Hmm maybe not for `ReportActionItemReimbursementQueued` since we have a button

nkuoch replied:
I don't think so because we're not using the fragmented message from the reportAction

luacmartins commented on src/pages/home/report/ReportActionItemReimbursementQueued.js:
```diff
@@ -0,0 +1,39 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import {View} from "react-native";
```
```suggestion
import {View} from 'react-native';
```

luacmartins commented on src/pages/home/report/ReportActionItemReimbursementQueued.js:
```diff
@@ -0,0 +1,39 @@
+import withLocalize, {withLocalizePropTypes} from '../../../components/withLocalize';
+import * as BankAccounts from '../../../libs/actions/BankAccounts';
+import * as store from '../../../libs/actions/ReimbursementAccount/store';
+import Button from "../../../components/Button";
```
```suggestion
import Button from '../../../components/Button';
```

luacmartins commented on tests/unit/ReportUtilsTest.js:
```diff
@@ -291,81 +291,100 @@ describe('ReportUtils', () => {
                 ...LHNTestUtils.getFakeReport(),
-                iouReportID: '1',
-                hasOutstandingIOU: true,
+                iouReportID: '1',q
```
```suggestion
                iouReportID: '1',
```

luacmartins reviewed (changes requested):
Looking good. Left a few comments.

nkuoch committed:
No need to show green dot on parent report

nkuoch commented on src/pages/ReimbursementAccount/CompanyStep.js:
```diff
@@ -59,10 +59,6 @@ class CompanyStep extends React.Component {
         this.defaultWebsite = lodashGet(props, 'user.isFromPublicDomain', false) ? 'https://' : `https://www.${Str.extractEmailDomain(props.session.email, '')}`;
     }
 
-    componentWillUnmount() {
-        BankAccounts.resetReimbursementAccount();
-    }
-
```
Just found it while I was testing adding a vbba. Errors didn't show up on this step because of it. We used to do it for each step but when refactoring with the new forms, this one got left, but there is no reason to do it for this step more than for the other steps.

nkuoch commented on src/libs/ReportUtils.js:
```diff
@@ -1874,44 +1897,54 @@ function isUnreadWithMention(report) {
     }
 
-    return report.hasOutstandingIOU;
+    if (reportToLook.ownerAccountID !== currentUserAccountID && reportToLook.hasOutstandingIOU) {
```
hmm I don't see why?

luacmartins replied:
We've been using `hasOutstandingIOU` as a way to tell if the report is settled or not, but we have report states for that. It'd be nice if we could standardize in one key for it and eventually remove the other.

nkuoch replied:
but now, `hasOutstandingIOU` doesn't necesserily mean `!isSettled`. Because when `isWaitingForBankAccount`, both are actually false. But I see your point. I'll try to think of a way to uniformize everything.

nkuoch commented on src/libs/ReportUtils.js:
```diff
@@ -236,7 +236,13 @@ function canFlagReportAction(reportAction) {
 function isSettled(reportID) {
-    return !lodashGet(allReports, [`${ONYXKEYS.COLLECTION.REPORT}${reportID}`, 'hasOutstandingIOU']);
+    const report = lodashGet(allReports, `${ONYXKEYS.COLLECTION.REPORT}${reportID}`, {});
+    return !report.hasOutstandingIOU && !report.isWaitingOnBankAccount;
```
No, we don't want to consider it as settled when isWaitingOnBankAccount is true, in which case, state is actually 3 (BILLING) and status is 3 (APPROVED). We could have used status but it isn't available in this scope yet.

Note: [Web](https://github.com/Expensify/Web-Expensify/pull/38042) and [Auth](https://github.com/Expensify/Auth/pull/8226) PRs related to that.

nkuoch commented on src/libs/ReportUtils.js:
```diff
@@ -1874,44 +1897,54 @@ function isUnreadWithMention(report) {
-    if (!report || !report.iouReportID || _.isUndefined(report.hasOutstandingIOU)) {
-        return false;
+function isWaitingForIOUActionFromCurrentUser(report) {
+    let reportToLook = report;
```
because if I don't:

<img width="793" alt="image" src="https://github.com/Expensify/App/assets/2463975/17438803-f8c5-41f2-9923-09354d5de42a">

nkuoch commented on src/libs/ReportUtils.js:
```diff
@@ -1874,44 +1897,54 @@ function isUnreadWithMention(report) {
-    if (report.hasOutstandingIOU) {
-        const iouReport = iouReports[`${ONYXKEYS.COLLECTION.REPORT}${report.iouReportID}`];
+function isIOUOwnedByCurrentUser(report) {
+    let reportToLook = report;
```
Same as above

nkuoch commented on src/libs/SidebarUtils.js:
```diff
@@ -148,31 +148,38 @@ function getOrderedReportIDs(currentReportId) {
 
-        if (report.hasOutstandingIOU && !ReportUtils.isIOUOwnedByCurrentUser(report, allReports)) {
+        const isIOUOwnedByCurrentUser = ReportUtils.isIOUOwnedByCurrentUser(report, allReports);
+        if (report.hasOutstandingIOU && !isIOUOwnedByCurrentUser) {
```
No it's not the same logic - isSettled doesn't take into account the current user

luacmartins replied:
Sorry, I meant instead of `report.hasOutstandingIOU`

nkuoch committed:
Small fixes from Carlos PR review

nkuoch committed:
Merge branch 'main' into nat-iouw

nkuoch commented on src/libs/ReportUtils.js:
```diff
@@ -1874,44 +1897,54 @@ function isUnreadWithMention(report) {
+function isWaitingForIOUActionFromCurrentUser(report) {
+    let reportToLook = report;
+
+    if (report.iouReportID) {
```
I'm not calling this method anymore for the green dot logic. Just for the logic of whether we need to display the report on the left panel or not.

But yes, the idea is:
- If it's the chat report, check if its child iou report has an action to do for the current user
- If it's the iou report, check  if it has an action to do for the current user

This way, we're sure both are displayed on the left.
Green dot does show on both reports (chat and iou) for the payer when they need to pay (because they have a button to pay on either).
But yes, for the waitingForBankAccount case, the green dot only shows on the iou report.

nkuoch committed:
Fix lint

nkuoch committed:
js prettier

nkuoch commented:
TODO: test doing a new request after a previous payment ... I had a weird issue where my new requests were automatically paid ... will see if I can reproduce tomorrow.

nkuoch committed:
Move addbankaccount button to MoneyRequestHeader and create generic ReportActionItemBasicMessage instead of ReportActionItemReimbursed and ReportActionItemReimbursementQueued

nkuoch committed:
Fix tests and style

nkuoch commented:
TODO: We should prevent adding request to pending one.

nkuoch committed:
If current user already has a iou waitingForBankAccount, prevent reusing same iouReportID for other requests

nkuoch committed:
Some refactoring

nkuoch committed:
Merge branch 'main' into nat-iouw

nkuoch committed:
Show green dot when admin needs to pay his own request

nkuoch committed:
Add button back to reportAction level

nkuoch committed:
Merge branch 'main' into nat-iouw

nkuoch committed:
Fix test

nkuoch committed:
prettier

nkuoch committed:
Fix tests

nkuoch committed:
Rename method to hasIouWaitingOnCurrentUserBankAccount

nkuoch committed:
Prettier

nkuoch committed:
Small fixes

nkuoch committed:
Prettier

luacmartins commented on src/pages/home/report/ReportActionItemBasicMessage.js:
```diff
@@ -0,0 +1,29 @@
+    children: null,
+};
+
+function ReportActionItemBasicMessage(props) {
```
Not a fan of this name as it doesn't seem specific enough, but I'm also having trouble coming up with a better suggestion 😅

nkuoch replied:
yeah well, it's supposed to be a generic name, so we can easily pass just a message for any kind of action.

luacmartins reviewed (commented):
PR is looking good, but I still haven't tested it. We also have conflicts and jest tests are failing.

mountiny reviewed (dismissed):
1. When Clicking the `Pay With Expensify` the request has been optimistically paid. Then in a split of a second it changed to the pending pattern waiting to add the bank account while the header of the expense report blinked disappearing. See point 6 for a similar video


2. After that the expense report title in the chat was stack on Workspace paid $10.00 - Pending 

<img width="443" alt="image" src="https://github.com/Expensify/App/assets/36083550/557e8a55-06b5-45fe-a998-29c1b07cc65e">

3. As an employee you can see the green dot to add bank account also on the workspace chat in the LHN but from this chat it gives you no indication of what should be the next steps or what you have to do there, you need to click into the expense report. This might be because we cannot load the expense reports on OpenApp for now, but once we can do it, we should probably have the dot only at the expense report which is held

<img width="1255" alt="image" src="https://github.com/Expensify/App/assets/36083550/c10800e2-77c1-4c14-904c-9f1956e7ce7e">

4. When adding the personal bank account, I feel like this Continue button at the end should close the RHP modal and go back to the reportID. We should use the Navigation.dismissModal(expenseReportID) method there to replace the RHP from the stack. this will clearly indicate to the user everything is done and they can also see the payment has been successful as now when we dont close the RHP they cant see its completed 
<img width="1239" alt="image" src="https://github.com/Expensify/App/assets/36083550/d0e1d689-8238-4455-827b-0e8d0a89c119">

5. The admin did not get the update about the settlement via pusher, I had to refresh the page to see the report has been paid and settled actually

6. Similarly to the first one, when admin pays from the report preview component, it first says it was paid, then the total report value changes to $0 and then finally reverts to original. However the policy expense chat does not have any indication of the user not having the bank account. Here is the video:

https://github.com/Expensify/App/assets/36083550/1345128b-260e-4f20-9a79-8bbacc806651



I think these are all quite minor but we should consider them. Especially the one where it first optimistically shows the payment has been successful and title of an expense report is showing in correctly as a result.

nkuoch committed:
Merge branch 'main' into nat-iouw

nkuoch committed:
Dont show green button on chat report for isWaitingOnBankAccount case

nkuoch committed:
Dont show green button on chat report for isWaitingOnBankAccount case

nkuoch committed:
Default report preview amount to the one from the report action if iouReport is not available

nkuoch committed:
Default report preview amount to the one from the report action if iouReport is not available

nkuoch committed:
Make sure we close the modal after clicking on Continue

mountiny commented:
There seem to be some lint failures, let us know when you think this is ready for another round of testing

nkuoch committed:
Merge branch 'main' into nat-iouw

nkuoch committed:
Fix missing iou amount

nkuoch committed:
Fix merge conflict

nkuoch committed:
Merge branch 'nat-iouw' of github.com:Expensify/App into nat-iouw

nkuoch commented:
Still on hold. Couldn't properly test again because of the "Error: Provided collection doesn't have all its data belonging to the same parent" error everyone is getting.

nkuoch committed:
prettier

mountiny commented:
We are working on fixing that one

nkuoch committed:
Merge branch 'vit-bumpOnyxVersion' into nat-iouw

nkuoch committed:
Merge branch 'main' into nat-iouw

mountiny commented:
This should be fixed on main now

nkuoch committed:
shouldShowGreenDotIndicator needs managerID in optionItem

nkuoch committed:
Fix unit test

nkuoch committed:
Fix ioupreview amount for other types of messages

nkuoch committed:
Fix preview message

nkuoch committed:
Close modal to show report only if coming from click from button from report

nkuoch commented:
Latest testing video:
https://github.com/Expensify/App/assets/2463975/56f47a23-9cf3-4358-be34-468108d60b07

> 1. and 6 (blinking)

Kind of Fixed (not going back to $0 in the preview in between, but still have the checkmark first before disappearing). But the blinking is still occurring on the left panel - Which will probably be much better once iouReports are always properly returned from auth (https://github.com/Expensify/Expensify/issues/278883#top).

> 2. After that the expense report title in the chat was stack on Workspace paid $10.00 - Pending

Is there any issue with that?

> 3. we should probably have the dot only at the expense report which is held

Fixed

> 4. this Continue button at the end should close the RHP modal and go back to the reportID

Fixed

> 5. The admin did not get the update about the settlement via pusher, I had to refresh the page to see the report has been paid and settled actually

Fixed via [Web PR](https://github.com/Expensify/Web-Expensify/pull/38172)

nkuoch committed:
Fix eslint

nkuoch committed:
If no iouReport to get managerID from, get actorAccountID from reportAction

nkuoch commented:
If there is no blocker, I suggest we merge this PR as is, and tackle any minor issues in follow up PRs as this one is starting to be biiig. In any cases, this PR is already fixing several existing bugs, so that's a win.

luacmartins commented:
@eVoloshchak all holding PRs are in production. You can test/review this one.

luacmartins commented on src/components/ReportActionItem/ReportPreview.js:
```diff
@@ -84,11 +88,27 @@ const defaultProps = {
+        reportAmount = '';
+        const actionMessage = _.size(props.action.message) ? props.action.message[0].text || '' : '';
+        const splits = actionMessage.split(' ');
+        for (let i = 0; i < splits.length; i++) {
+            if (/\d/.test(splits[i])) {
+                reportAmount = splits[i];
+            }
+        }
```
What if the user adds a number in the comment? That'd replace the amount, wouldn't it? I think we could achieve the same result without the for loop, by using the first match in this regex instead:

```
\S+\d[\d,\.]*?\b
```

nkuoch replied:
I don't think the message in the preview action contains any user comment - And if you see the 3 examples the previous comment, this wouldn't work:
```
'Domain20821\'s Workspace owes $33.00'.match(/\S+\d[\d,\.]*?\b/)
['Domain20821', index: 0, input: "Domain20821's Workspace owes $33.00", groups: undefined]
```

mountiny replied:
This feels like a temporary solution for sure, its probably quite brittle with different workspace names and also different currencies, I think we need to ensure any report preview which is shown to the user has the associated iou report loaded in the app. But that can be a polish

luacmartins commented on src/components/ReportActionItem/ReportPreview.js:
```diff
@@ -123,7 +141,7 @@ function ReportPreview(props) {
                         </View>
                     </View>
-                    {isCurrentUserManager && !ReportUtils.isSettled(props.iouReport.reportID) && (
+                    {props.iouReport && isCurrentUserManager && !ReportUtils.isSettled(props.iouReport.reportID) && !props.iouReport.isWaitingOnBankAccount && (
```
We have `{}` as a default value for the iouReport prop in line 83, so `props.iouReport` is always truthy

```suggestion
                    {isCurrentUserManager && !ReportUtils.isSettled(props.iouReport.reportID) && !props.iouReport.isWaitingOnBankAccount && (
```

luacmartins commented on src/libs/ReportUtils.js:
```diff
@@ -229,7 +229,13 @@ function canEditReportAction(reportAction) {
 function isSettled(reportID) {
-    return !lodashGet(allReports, [`${ONYXKEYS.COLLECTION.REPORT}${reportID}`, 'hasOutstandingIOU']);
+    const report = lodashGet(allReports, `${ONYXKEYS.COLLECTION.REPORT}${reportID}`, {});
+    return report && !report.hasOutstandingIOU && !report.isWaitingOnBankAccount;
```
Report will default to {} which is always truthy

```suggestion
    return !report.hasOutstandingIOU && !report.isWaitingOnBankAccount;
```

luacmartins commented on src/libs/ReportUtils.js:
```diff
@@ -229,7 +229,13 @@ function canEditReportAction(reportAction) {
+    return report && !report.hasOutstandingIOU && !report.isWaitingOnBankAccount;
+}
+
+function isCurrentUserSubmitter(reportID) {
```
Missing JSDocs

luacmartins commented on src/libs/ReportUtils.js:
```diff
@@ -2299,13 +2336,28 @@ function getReportIDFromLink(url) {
     return reportID;
 }
 
+function hasIouWaitingOnCurrentUserBankAccount(chatReport) {
```
Missing JSDocs

```suggestion
function hasIOUWaitingOnCurrentUserBankAccount(chatReport) {
```

luacmartins commented on src/libs/actions/BankAccounts.js:
```diff
@@ -33,8 +33,13 @@ function openPlaidView() {
 
-function openPersonalBankAccountSetupView() {
-    clearPlaid().then(() => Navigation.navigate(ROUTES.SETTINGS_ADD_BANK_ACCOUNT));
+function openPersonalBankAccountSetupView(exitReportID) {
```
Missing JSDocs

luacmartins reviewed (dismissed):
Left a few comments. While testing, I noticed a few things:

1. We display `paid -$X using ACH` for a brief second before showing `started settling up...`
2. Once the employee adds a bank account, it seems that the admin report is not updated. I didn't see the `x added a bank account...` message and the request still shows as pending.

https://github.com/Expensify/App/assets/22219519/99d43da5-f2db-4107-9107-8fdd72fc278c

nkuoch committed:
Update src/components/ReportActionItem/ReportPreview.js

Co-authored-by: Carlos Martins <cmartins@expensify.com>

nkuoch committed:
Fixes from Carlos PR review

nkuoch commented:
> Once the employee adds a bank account, it seems that the admin report is not updated. I didn't see the x added a bank account... message and the request still shows as pending.

It's because you need to pull Web branch https://github.com/Expensify/Web-Expensify/pull/38172

nkuoch commented:
> We display paid -$X using ACH for a brief second before showing started settling up...

Yeah ... this will need to be addressed in a follow up issue, as there is no easy fast solution for that (https://expensify.slack.com/archives/C02HWMSMZEC/p1689056981435559?thread_ts=1689033812.851139&cid=C02HWMSMZEC)

nkuoch commented:
Created [new issue](https://github.com/Expensify/Expensify/issues/300006) to address polishing issues

nkuoch commented:
Will be OOO until wednesday - don't hesitate to push yourself in the PR if needed meanwhile.

nkuoch committed:
Dont show settlement button if we dont have the ioureport

mountiny commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

</details>

mountiny commented on src/libs/ReportUtils.js:
```diff
@@ -2299,13 +2342,34 @@ function getReportIDFromLink(url) {
  * @returns {Boolean}
  */
 function canRequestMoney(report) {
+    // Prevent requesting money if pending iou waiting for their bank account already exists.
+    if (hasIOUWaitingOnCurrentUserBankAccount(report)) {
+        return false;
+    }
```
I think we need to add this same check to the global create as now the user can request money from the workspace using the global create

mountiny replied:
added

mountiny commented on src/libs/actions/BankAccounts.js:
```diff
@@ -33,8 +33,17 @@ function openPlaidView() {
+        if (exitReportID) {
+            Onyx.merge(ONYXKEYS.PERSONAL_BANK_ACCOUNT, {exitReportID});
+        }
+        Navigation.navigate(ROUTES.SETTINGS_ADD_BANK_ACCOUNT);
```
I think we should use dismissModal as it might help in case the navigates back to the plaid, But NAB

mountiny reviewed (commented):
Tests well, doing partial review now.

mountiny committed:
Global create update

mountiny committed:
Resolve merge conflicts

luacmartins commented:
@mountiny lint and jest are failing.

mountiny committed:
Merge branch 'main' into nat-iouw

mountiny committed:
Fix linter

mountiny commented:
Updated

luacmartins commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

</details>

mountiny committed:
Resolve merge conflicts

mountiny committed:
Update tests

luacmartins commented:
We still have a failing test and I'm getting the following error when paying the request as the admin:

![Screenshot 2023-07-17 at 11 46 44 AM](https://github.com/Expensify/App/assets/22219519/6952f4de-67b3-4d85-bea4-aed0fb717ece)

mountiny committed:
Update the tests

luacmartins commented:
@mountiny @nkuoch I'm still getting the error above when paying the request.

mountiny committed:
Resolve merge conflicts

mountiny committed:
Remove duplicate single transaction details

mountiny commented:
I havent been able to figure this one in the time which left to me, I have resolved some conflicts at least.

@nkuoch How do I run locally the REIMBURSEMENTQUEUED logic ie I have requested the money and admin paid but I need to addd the bank account. What do I need to do so this logic runs when I add the bank account? Or I have added it with script, do I have to add it from the product?

nkuoch committed:
Merge branch 'main' into nat-iouw

nkuoch committed:
Fix merge conflict

nkuoch committed:
Fix cannot read property of null

nkuoch committed:
Revert console.log

nkuoch commented:
> How do I run locally the REIMBURSEMENTQUEUED logic ie I have requested the money and admin paid but I need to addd the bank account. What do I need to do so this logic runs when I add the bank account? Or I have added it with script, do I have to add it from the product?

Yes add it in product, clicking on the button shown to the requestor, adding Plaid test bank account (user_good, pass_good)

nkuoch committed:
Import underscore

nkuoch committed:
Merge branch 'main' into nat-iouw

nkuoch committed:
Fix merge conflict

nkuoch committed:
Fix eslint

luacmartins reviewed (changes requested):
I tested the flow again today and I'm seeing two bugs that I think are blockers:

**BUG 1: Expense report disappears from LHN and ReportScreen**
1. As the admin, pay the money request
2. Notice that the report temporarily disappears from the LHN and ReportScreen and we're navigated to the Concierge chat

**BUG 2: Settlement button is still visible after paying and crashes the App**
1. After the steps above, navigate back to the Expense report once it reappears in the LHN
4. Notice that the settlement button is still visible (it shouldn't)
5. Tap the button
6. Notice that the App crashes

https://github.com/Expensify/App/assets/22219519/9d94c41e-0d52-4ea9-a1ac-db16a3839bb4

**BUG 3: Settlement is not triggered when employee adds bank account**
1. After the payment was initiated by the admin, as the employee, add your bank account 
2. Notice that the `added a bank account...` message doesn't show up and the payment is not automatically triggered.

https://github.com/Expensify/App/assets/22219519/768ea665-9ea2-4a81-8cdf-271d90069da1

Let me know if I'm missing something.

nkuoch committed:
Merge branch 'main' into nat-iouw

nkuoch committed:
Fix addBankAccount button still visible after user adds a bank account

nkuoch committed:
Only show iou.pay, not reimbursed action to avoid duplicate message

luacmartins commented:
Thanks for addressing those @nkuoch. Retesting this now.

luacmartins commented:
I started a group DM so we can get this merged faster, but I'm still seeing some issues when the employee adds a bank account:

https://github.com/Expensify/App/assets/22219519/e6946dbf-ade2-448e-abd3-cf66289fef51

luacmartins reviewed (commented):
I started a group DM so we can get this merged faster, but I'm still seeing some issues when the employee adds a bank account:

https://github.com/Expensify/App/assets/22219519/e6946dbf-ade2-448e-abd3-cf66289fef51

nkuoch committed:
Fix test

luacmartins reviewed (approved):
LGTM and tests well. 

Since @mountiny had already reviewed this PR but then committed to the branch and C+ won't be able to reliably test this flow given the restricted access to bank accounts, I'm gonna go ahead and merge this PR. We'll address the smaller issues in follow up PRs.

luacmartins commented:
Tests passed, removing label

![Screenshot 2023-07-25 at 2 09 09 PM](https://github.com/Expensify/App/assets/22219519/3d0c22f8-713d-4063-b326-be1b396a0912)

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5670777418) to staging by https://github.com/luacmartins in version: 1.3.46-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5685630056) to staging by https://github.com/luacmartins in version: 1.3.47-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5685624741) to production by https://github.com/marcaaron in version: 1.3.46-2 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

eVoloshchak commented on src/pages/AddPersonalBankAccountPage.js:
```diff
@@ -92,17 +106,15 @@ class AddPersonalBankAccountPage extends React.Component {
-                        onButtonPress={() => {
-                            Navigation.navigate(ROUTES.SETTINGS_PAYMENTS);
-                        }}
+                        onButtonPress={this.exitFlow}
```
👋 Coming from https://github.com/Expensify/App/issues/38558, this has caused a small issue where the same page was added twice to the navigation stack
